src/engine/Settings.cs

```diff
@@ -64,74 +112,153 @@ private Settings()
+    public bool RunAutoEvoDuringGamePlay { get; set; } = true;
 
-    /// <summary>
-    ///   If this is over 0 then this limits how often compound clouds
-    ///   are updated. The default value of 0.020 at 60 FPS makes
-    ///   every other frame not update the clouds.
-    /// </summary>
-    /// <remarks>
-    ///   <para>
-    ///     This should be made user configurable for different
-    ///     computers. The choises should probably be:
-    ///     0.0f, 0.020f, 0.040f, 0.1f, 0.25f
-    ///   </para>
-    /// </remarks>
-    public float CloudUpdateInterval { get; set; } = 0.040f;
+    public int CloudSimulationWidth => Constants.CLOUD_X_EXTENT / CloudResolution;
+
+    public int CloudSimulationHeight => Constants.CLOUD_Y_EXTENT / CloudResolution;
+
+    public static bool operator ==(Settings lhs, Settings rhs)
+    {
+        return Equals(lhs, rhs);
+    }
+
+    public static bool operator !=(Settings lhs, Settings rhs)
+    {
+        return !(lhs == rhs);
+    }
+
+    public override bool Equals(object obj)
+    {
+        if (obj == null)
+        {
+            return false;
+        }
+
+        if (GetType() != obj.GetType())
+        {
+            return false;
+        }
+
+        return Equals((Settings)obj);
+    }
+
+    public bool Equals(Settings obj)
+    {
+        // Compare all properties in the two objects for equality.
+        var type = GetType();
+
+        foreach (var property in type.GetProperties())
+        {
+            // Returns if any of the properties don't match.
+            object thisValue = property.GetValue(this);
+            object objValue = property.GetValue(obj);
+
+            if (thisValue != objValue && (thisValue == null || !thisValue.Equals(objValue)))
+            {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    public override int GetHashCode()
+    {
+        int hashCode = 17;
+
+        var type = GetType();
+
+        foreach (var property in type.GetProperties())
+        {
+            int nextHash = property.GetHashCode();
+
+            hashCode ^= nextHash;
+        }
+
+        return hashCode;
+    }
 
     /// <summary>
-    ///   Sets amount of MSAA to apply to the viewport
+    ///   Returns a cloned deep copy of the settings object.
     /// </summary>
-    public Viewport.MSAA MSAAResolution { get; set; } = Viewport.MSAA.Disabled;
+    public Settings Clone()
+    {
+        Settings settings = new Settings();
+        settings.CopySettings(this);
+
+        return settings;
+    }
 
     /// <summary>
-    ///   Choose what filter to apply to the screen
-    ///   0 = None, 1 = Red/Green, 2 = Blue/Yellow
+    ///   Loads values from the saved setting configuration file.
     /// </summary>
-    public int ColourblindSetting { get; set; } = 0;
+    /// <returns>True on success, false if the file is invalid or couldn't be opened for reading.</returns>
+    public bool LoadFromFile()
```

Comment:
Is there a need to have this method separately?
I liked the previous load method (`LoadSettingsFileOrDefault()`) that didn't create an extra Settings object just to copy things from it.

Reply:
I could, currently it wouldn't make much difference. I just wrote it that way and left it in case we needed to load settings from the config file at some point. If I revert it back to how it was before then if you need to at any point you'll have to create it again, since you can't just reassign the singleton with it being readonly.

Reply:
I don't see that really happening. If we were to add folder change listening and reload the settings then if the json file is changed on disk, then I could see it being useful. But I at least would like the load from json to directly return the Settings object it loaded. That way if later it is needed it is just 2 lines to make a method to load and copy settings from json to an existing Settings object.